src/Controllers/Common/ContentController.cs

```diff
@@ -2293,4 +2194,106 @@ public IActionResult GetRevealIndex()
+                // resolve items in the bundle
+                foreach (var reward in bundleItem.Relationship.Where(e => e.Type == "Bundle")) {
+                    int quantity = itemService.GetBulkItemQuantity(reward, i.Value);
+                    inventoryItemsToAdd.TryAdd(reward.ItemId, 0);
+                    inventoryItemsToAdd[reward.ItemId] += quantity;
+                    itemsToSendBack.TryAdd(reward.ItemId, 0);
+                    itemsToSendBack[reward.ItemId] += quantity;
+                }
+            }
+            else if (itemService.IsBoxItem(i.Key)) {
+                // open boxes individually
+                for (int j = 0; j < i.Value; j++) {
+                    itemService.OpenBox(i.Key, gender, out int itemId, out int quantity);
+                    inventoryItemsToAdd.TryAdd(itemId, 0);
+                    inventoryItemsToAdd[itemId] += quantity;
+                    itemsToSendBack.TryAdd(itemId, 0);
+                    itemsToSendBack[itemId] += quantity;
+                }
+            }
+            else {
+                // add item to inventory
+                inventoryItemsToAdd.TryAdd(i.Key, 0);
+                inventoryItemsToAdd[i.Key] += i.Value;
+                itemsToSendBack.TryAdd(i.Key, 0);
+                itemsToSendBack[i.Key] += i.Value;
+            }
+        }
+
+        // check if the user can afford the purchase
+        if (currency.GameCurrency - totalCoinCost < 0 && currency.CashCurrency - totalGemCost < 0) {
+            return new CommonInventoryResponse {
+                Success = false,
+                CommonInventoryIDs = new CommonInventoryResponseItem[0],
+                UserGameCurrency = achievementService.GetUserCurrency(viking)
+            };
+        }
+
+        // deduct the cost of the purchase
+        achievementService.AddAchievementPoints(viking, AchievementPointTypes.GameCurrency, -totalCoinCost + gemsToAdd);
+        achievementService.AddAchievementPoints(viking, AchievementPointTypes.CashCurrency, -totalGemCost + coinsToAdd);
+
+        // add items to the inventory (database)
+        var addedItems = inventoryService.AddItemsToInventoryBulk(viking, inventoryItemsToAdd);
+
+        // build response
+        List<CommonInventoryResponseItem> items = new List<CommonInventoryResponseItem>();
+        foreach (var i in itemsToSendBack) {
+            items.AddRange(Enumerable.Repeat(
+                new CommonInventoryResponseItem {
+                    CommonInventoryID = addedItems.ContainsKey(i.Key) ? addedItems[i.Key] : 0, // return inventory id if this item was added to the DB
+                    ItemID = i.Key,
+                    Quantity = 0
+                }, i.Value));
+        }
+        // NOTE: The quantity of purchased items is always 0 and the items are instead duplicated in both the request and the response.
+        //       Call AddItemToInventoryAndGetResponse with Quantity == 1 we get response with quantity == 0.
```

Comment:
this note need be updated ... no more `AddItemToInventoryAndGetResponse` here but still we are need to use not obvious `Quantity = 0` and repeat `CommonInventoryResponseItem` for the same item ...

Reply:
Adjusted